minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/service/MiNiFiConfigurationChangeListener.java

```diff
@@ -0,0 +1,208 @@
+            persistBackNonFlowSectionsFromOriginalSchema(bufferedConfigOs.toByteArray(), bootstrapProperties, configFile);
+
+            // Create an input stream to feed to the config transformer
+            try (FileInputStream newConfigIs = new FileInputStream(configFile)) {
+
+                try {
+                    String confDir = bootstrapProperties.getProperty(CONF_DIR_KEY);
+
+                    try {
+                        logger.info("Performing transformation for input and saving outputs to {}", confDir);
+                        ByteBuffer tempConfigFile = generateConfigFiles(newConfigIs, confDir, bootstrapFileProvider.getBootstrapProperties());
+                        runner.getConfigFileReference().set(tempConfigFile.asReadOnlyBuffer());
+
+                        try {
+                            logger.info("Reloading instance with new configuration");
+                            restartInstance();
+                        } catch (Exception e) {
+                            logger.debug("Transformation of new config file failed after transformation into Flow.xml and nifi.properties, reverting.");
+                            try (FileInputStream swapConfigFileStream = new FileInputStream(swapConfigFile)) {
+                                ByteBuffer resetConfigFile = generateConfigFiles(swapConfigFileStream, confDir, bootstrapFileProvider.getBootstrapProperties());
+                                runner.getConfigFileReference().set(resetConfigFile.asReadOnlyBuffer());
+                            }
+                            throw e;
+                        }
+                    } catch (Exception e) {
+                        logger.debug("Transformation of new config file failed after replacing original with the swap file, reverting.");
+                        try (FileInputStream swapConfigFileStream = new FileInputStream(swapConfigFile)) {
+                            Files.copy(swapConfigFileStream, configFile.toPath(), REPLACE_EXISTING);
+                        }
+                        throw e;
+                    }
+                } catch (Exception e) {
+                    logger.debug("Transformation of new config file failed after swap file was created, deleting it.");
+                    if (!swapConfigFile.delete()) {
+                        logger.warn("The swap file failed to delete after a failed handling of a change. It should be cleaned up manually.");
+                    }
+                    throw e;
+                }
+            }
+        } catch (ConfigurationChangeException e){
+            logger.error("Unable to carry out reloading of configuration on receipt of notification event", e);
+            throw e;
+        } catch (IOException ioe) {
+            logger.error("Unable to carry out reloading of configuration on receipt of notification event", ioe);
+            throw new ConfigurationChangeException("Unable to perform reload of received configuration change", ioe);
```

Comment:
Is there a reason for logging the error and throw the exception? Shouldn't the exception be caught and logged by the caller? This approach seems like it would generate multiple log messages for the same problem.

Reply:
The ConfigurationChangeCoordinator logs the ListenerHandlerResult so the result here will be logged. I removed the error logging to avoid duplication.